src/Symfony/Component/Messenger/Transport/Doctrine/Connection.php

```diff
@@ -0,0 +1,224 @@
+            'auto_setup' => $options['auto_setup'] ?? ($query['auto_setup'] ?? true),
+        ];
+    }
+
+    public function publish(string $body, array $headers): void
+    {
+        $sql = $this->driverConnection->createQueryBuilder()
+            ->insert($this->configuration['table_name'])
+            ->values([
+                'body' => ':body',
+                'headers' => ':headers',
+                'queue' => ':queue',
+                'created_at' => ':now',
+                'available_at' => ':now',
+            ]);
+        $now = (new \DateTime('now'))->format('Y-m-d H:i:s');
+        $this->executeQuery($sql, [
+            ':body' => $body,
+            ':headers' => \json_encode($headers),
+            ':queue' => $this->configuration['queue'],
+            ':now' => $now,
+        ]);
+    }
+
+    public function get(): ?array
+    {
+        $this->driverConnection->beginTransaction();
+        try {
+            $query = $this->driverConnection->createQueryBuilder()
+                ->select('m.*')
+                ->from($this->configuration['table_name'], 'm')
+                ->where('m.delivered_at is null')
+                ->andWhere('m.available_at <= :now')
+                ->andWhere('m.queue = :queue')
+                ->orderBy('available_at', 'ASC')
+                ->setMaxResults(1);
+
+            // use SELECT ... FOR UPDATE to lock table
+            $doctrineEnvelope = $this->executeQuery(
+                $query->getSQL().' '.$this->driverConnection->getDatabasePlatform()->getWriteLockSQL(),
+                [
+                    ':now' => (new \DateTime('now'))->format('Y-m-d H:i:s'),
+                    ':queue' => $this->configuration['queue'],
+                ]
+            )->fetch();
+
+            if (false === $doctrineEnvelope) {
+                $this->driverConnection->commit();
+
+                return null;
+            }
+
+            $doctrineEnvelope['headers'] = \json_decode($doctrineEnvelope['headers'], true);
+
+            $sql = $this->driverConnection->createQueryBuilder()
+                ->update($this->configuration['table_name'])
+                ->set('delivered_at', ':delivered_at')
+                ->where('id = :id');
+            $this->executeQuery($sql, [
+                ':id' => $doctrineEnvelope['id'],
+                ':delivered_at' => (new \DateTime('now'))->format('Y-m-d H:i:s'),
+            ]);
+
+            $this->driverConnection->commit();
+
+            return $doctrineEnvelope;
+        } catch (\Throwable $e) {
+            $this->driverConnection->rollBack();
+
+            throw $e;
+        }
+    }
+
+    public function ack($id): bool
+    {
+        return $this->driverConnection->delete($this->configuration['table_name'], ['id' => $id]) > 0;
+    }
+
+    public function reject($id): bool
+    {
+        return $this->driverConnection->update($this->configuration['table_name'], ['delivered_at' => null], ['id' => $id]) > 0;
```

Comment:
This needs to be a delete. This happens in 2 cases:

1) Message decoding fails
2) We've retried a few times, and we're giving up

In both cases, this message needs to be permanently gone. Right now, after it retries a few times, it stay there, and the worker keeps trying it again forever :) 

I know, it's weird that `ack()` and `reject()` both do the same thing. But... it *is* how it's supposed to work :).

> Positive acknowledgements simply instruct RabbitMQ to record a message as delivered and can be discarded. Negative acknowledgements with basic.reject have the same effect. The difference is primarily in the semantics: positive acknowledgements assume a message was successfully processed while their negative counterpart suggests that a delivery wasn't processed but still should be deleted.

https://www.rabbitmq.com/confirms.html